src/core/provider/MCAccountValidator.cpp

```diff
@@ -0,0 +1,421 @@
+        resolveMXDone();
+    }
+    else {
+        checkNextHostDone();
+    }
+}
+
+void AccountValidator::resolveMX()
+{
+    Array * components;
+    String * domain;
+    
+    components = mUsername->componentsSeparatedByString(MCSTR("@"));
+    if (components->count() >= 2) {
+        domain = (String *) components->lastObject();
+        mResolveMX = new MXRecordResolverOperation();
+        mResolveMX->setHostname(domain);
+        mResolveMX->setCallback((OperationCallback *)this);
+        
+        mQueue = new OperationQueue();
+        mQueue->addOperation(mResolveMX);
+    }
+    else {
+        mImapError = ErrorNoValidServerFound;
+        mPopError = ErrorNoValidServerFound;
+        mSmtpError = ErrorNoValidServerFound;
+        
+        callback()->operationFinished(this);
+    }
+}
+
+void AccountValidator::resolveMXDone()
+{
+    Array * mxRecords = mResolveMX->mxRecords();
+    
+    mc_foreacharray(String, mxRecord, mxRecords) {
+        MailProvider *provider = MailProvidersManager::sharedManager()->providerForMX(mxRecord);
+        if (provider){
+            mProvider = provider;
+            break;
+        }
+    }
+    
+    startCheckingHosts();
+}
+
+void AccountValidator::startCheckingHosts()
+{
+    if (mProvider != NULL) {
+        mIdentifier = mProvider->identifier();
+        
+        if (mImapServices->count() == 0 and mProvider->imapServices()->count() > 0)
+            mImapServices = mProvider->imapServices();
+        
+        if (mPopServices->count() == 0 and mProvider->popServices()->count() > 0)
+            mPopServices = mProvider->popServices();
+        
+        if (mSmtpServices->count() == 0 and mProvider->smtpServices()->count() > 0)
+            mSmtpServices = mProvider->smtpServices();
+    }
+    
+    if (mImapServices->count() == 0)
+        mImapError = ErrorNoValidServerFound;
+    
+    if (mPopServices->count() == 0)
+        mPopError = ErrorNoValidServerFound;
+    
+    if (mSmtpServices->count() == 0)
+        mSmtpError = ErrorNoValidServerFound;
+    
+    checkNextHost();
+}
+
+/**
+ Each service(IMAP/POP/SMTP) is tested one after the other.
+ For each service we test each server details (NetService),
+ Until either:  
+    we find on that works and returns ErrorNone in checkNextHostDone().
+    we have gone trough the Array of NetService for that service and checkNextHost() is then called for the next service.
+ */
+void AccountValidator::checkNextHost()
+{
+    if (mCurrentServiceTested == SERVICE_IMAP) {
+        if (mCurrentServiceIndex < mImapServices->count()) {
+            
+            if (mImapSession == NULL) {
+                mImapSession = new IMAPAsyncSession();
+                mImapSession->setUsername(mUsername);
+                mImapSession->setPassword(mPassword);
+            }
+            
+            mImapServer = (NetService *) mImapServices->objectAtIndex(mCurrentServiceIndex);
+            mImapSession->setHostname(mImapServer->hostname());
```

Comment:
Create a new session each time you test. Sessions are not designed to be reusable with different servers.